src/Npgsql/PoolManager.cs

```diff
@@ -24,27 +24,20 @@ static class PoolManager
 
         internal static bool TryGetValue(string key, [NotNullWhen(true)] out ConnectorPool? pool)
         {
-            // Note that pools never get removed. _pools is strictly append-only.
             var nextSlot = _nextSlot;
             var pools = _pools;
             // There is a race condition between TryGetValue and ClearAll.
             // When the nextSlot has already been read as a non-zero value, but the pools have already been reset.
             // And in this case, we will iterate beyond the size of the array.
             // As a safeguard, we take a min length between the array and the nextSlot.
             var minNextSlot = Math.Min(nextSlot, pools.Length);
-            var sw = new SpinWait();
 
             // First scan the pools and do reference equality on the connection strings
             for (var i = 0; i < minNextSlot; i++)
             {
                 var cp = pools[i];
                 if (ReferenceEquals(cp.Key, key))
                 {
-                    // It's possible that this pool entry is currently being written: the connection string
-                    // component has already been written, but the pool component is just about to be. So we
-                    // loop on the pool until it's non-null
-                    while (Volatile.Read(ref cp.Pool) == null)
-                        sw.SpinOnce();
                     pool = cp.Pool;
```

Comment:
It's not a trusted line anymore since the assignment of tuple parts isn't an atomic operation and even if you will change the order in code it doesn't mean that JIT or the processor will keep it the same. It's possible to use volatile operation and have memory fences, but let's not complicate things and just use `ConcurrentDictionary<,>`.

Reply:
I'm not sure there is a race condition here - whenever a pool is added, it will be set in the array's  `_nextSlot` position. But it's only incremented after we've written both key and the pool there.

Reply:
You're right, just checked what `Interlocked` does (unfortunately, not mentioned directly in its docs). Since there's a full memory fence the assignments won't be visible for other threads.

Reply:
Yes, this is my understanding as well. All interlocked operations implicitly also have a full memory barrier.